src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                ConsumeNumberResult result = ConsumeIntegerDigits(ref data, ref i);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+                if (nextByte != '.' && nextByte != 'E' && nextByte != 'e')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitComponentNotFound, nextByte);
+                }
+            }
+
+            Debug.Assert(nextByte == '.' || nextByte == 'E' || nextByte == 'e');
+
+            if (nextByte == '.')
+            {
+                i++;
+                ConsumeNumberResult result = ConsumeDecimalDigits(ref data, ref i, nextByte);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+                if (nextByte != 'E' && nextByte != 'e')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitEValueNotFound, nextByte);
+                }
+            }
+
+            Debug.Assert(nextByte == 'E' || nextByte == 'e');
+            i++;
+
+            signResult = ConsumeSign(ref data, ref i, nextByte);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            i++;
+            ConsumeNumberResult resultExponent = ConsumeIntegerDigits(ref data, ref i);
+            if (resultExponent == ConsumeNumberResult.NeedMoreData)
+                return false;
+            if (resultExponent == ConsumeNumberResult.Success)
+                goto Done;
+            if (resultExponent == ConsumeNumberResult.OperationIncomplete)
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, nextByte);
+            }
+
+        Done:
+            ValueSpan = data.Slice(0, i);
+            consumed = i;
+            return true;
+        }
+
+        private ConsumeNumberResult ConsumeNegativeSign(ref ReadOnlySpan<byte> data, ref int i)
+        {
+            byte nextByte = data[i];
+
+            if (nextByte == '-')
+            {
+                i++;
+                if (i >= data.Length)
+                {
+                    if (IsLastSpan)
+                    {
+                        _lineBytePosition += i;
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                    }
+                    else
```

Comment:
Braces

Reply:
I agree that we should add braces to improve readability.

I initially had braces, but a lot of the larger methods have single line `if (condition)` and it was useful for me to see the entire method body at once. With braces, there was too much extra lines and it made it especially difficult when reasoning about various edge cases.

I suspect some of this logic will be iterated on (especially for performance) so I was leaning towards keeping it compact for now and make the change to improve readability/maintainability when the methods are more stable.

Thoughts?

Reply:
There are two different types of "braces" comments here.

The ones Steve made are condensed forms of "This chunk of code violates our style guidelines in that it mixes braced and unbraced blocks within the same if/[else if/...][else] waterfall, and as such it should not be merged".

My one meta comment was about adding braces even when they weren't required.

Steve's bracing comments need to be addressed.  Based on a verbal discussion that you and I had it seems like your bracing style is similar to mine; and my general comment to that would be that you should never count on a followup change in master.  Assume that a branch can be snapped at any point and this code might have to get maintained on said branch.  If you'd be restyling it in master that'll make an ugly merge later. It also would mean maintaining code that you (the author) consider to be less readable than you'd like.  So, to avoid rambling for an hour, I'll summarize as "add any braces now that you wish were there later" (and even if you don't wish they were there, add the ones Steve mentioned).